File: src/main/java/com/kingdom/util/computercardaction/SalvationComputerCardActionHandler.java
package com.kingdom.util.computercardaction;

import com.kingdom.model.Card;
import com.kingdom.model.CardAction;
import com.kingdom.model.Game;
import com.kingdom.model.Player;
import com.kingdom.model.computer.ComputerPlayer;
import com.kingdom.util.cardaction.CardActionHandler;

import java.util.ArrayList;
import java.util.List;

public class SalvationComputerCardActionHandler {
    public static void handleCardAction(CardAction cardAction, ComputerPlayer computer) {

        Game game = computer.getGame();
        Player player = computer.getPlayer();

        String cardName = cardAction.getCardName();

        if (cardName.equals("Alms")) {
            //todo don't use this card if only have good treasure cards in hand
            Card cardToTrash = computer.getLowestCostCard(cardAction.getCards());
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(cardToTrash.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Archbishop")) {
            //todo better logic to determine best choice
            String choice;
            if (player.getActions() == 0) {
                choice = "actions";
            } else {
                choice = "sins";
            }
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, null, choice, -1);
        } else if (cardName.equals("Assassin")) {
            //todo determine best attack card to trash
            Card attackToTrash = computer.getLowestCostCard(cardAction.getCards());
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(attackToTrash.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Baptistry")) {
            //todo determine best card
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(Card.COPPER_ID);
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Bell Tower")) {
            //todo determine best choice
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, null, "after", -1);
        } else if (cardName.equals("Catacombs")) {
            Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(cardToGain.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Edict")) {
            //todo need better logic to determine best choice
            Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(cardToGain.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Graverobber")) {
            if (cardAction.getType() == CardAction.TYPE_CHOOSE_CARDS) {
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, new ArrayList<Integer>(), null, null, -1);
            } else {
                String yesNoAnswer;
                if (cardAction.getCards().get(0).getCost() > 0 || computer.wantsCoppers()) {
                    yesNoAnswer = "yes";
                } else {
                    yesNoAnswer = "no";
                }
                CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, null, yesNoAnswer, null, -1);
            }
        } else if (cardName.equals("Mendicant")) {
            Card cardToGain = computer.getHighestCostCard(cardAction.getCards());
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(cardToGain.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        } else if (cardName.equals("Scriptorium")) {
            //todo determine best action to discard
            Card cardToDiscard = computer.getLowestCostCard(cardAction.getCards());
            List<Integer> cardIds = new ArrayList<Integer>();
            cardIds.add(cardToDiscard.getCardId());
            CardActionHandler.INSTANCE.handleSubmittedCardAction(game, player, cardIds, null, null, -1);
        }
    }
}

File: src/main/kotlin/com/kingdom/util/computercardaction/SalvationComputerCardActionHandler.kt
package com.kingdom.util.computercardaction

import com.kingdom.model.cards.Card
import com.kingdom.model.OldCardAction
import com.kingdom.model.computer.ComputerPlayer
import com.kingdom.util.cardaction.CardActionHandler
import java.util.*

object SalvationComputerCardActionHandler {
    fun handleCardAction(oldCardAction: OldCardAction, computer: ComputerPlayer) {

        val game = computer.game
        val player = computer.player

        val cardName = oldCardAction.cardName

        when (cardName) {
            "Alms" -> {
                //todo don't use this card if only have good treasure cards in hand
                val cardToTrash = computer.getLowestCostCard(oldCardAction.cards)
                val cardIds = ArrayList<Int>()
                cardIds.add(cardToTrash!!.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Archbishop" -> {
                //todo better logic to determine best choice
                val choice = if (player.actions == 0) {
                    "actions"
                } else {
                    "sins"
                }
                CardActionHandler.handleSubmittedCardAction(game, player, null!!, null, choice, -1)
            }
            "Assassin" -> {
                //todo determine best attack card to trash
                val attackToTrash = computer.getLowestCostCard(oldCardAction.cards)
                val cardIds = ArrayList<Int>()
                cardIds.add(attackToTrash!!.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Baptistry" -> {
                //todo determine best card
                val cardIds = ArrayList<Int>()
                cardIds.add(Card.COPPER_ID)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Bell Tower" -> //todo determine best choice
                CardActionHandler.handleSubmittedCardAction(game, player, null!!, null, "after", -1)
            "Catacombs" -> {
                val cardToGain = computer.getHighestCostCard(oldCardAction.cards)
                val cardIds = ArrayList<Int>()
                cardIds.add(cardToGain!!.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Edict" -> {
                //todo need better logic to determine best choice
                val cardToGain = computer.getHighestCostCard(oldCardAction.cards)
                val cardIds = ArrayList<Int>()
                cardIds.add(cardToGain!!.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Graverobber" -> when {
                oldCardAction.type == OldCardAction.TYPE_CHOOSE_CARDS -> CardActionHandler.handleSubmittedCardAction(game, player, ArrayList(), null, null, -1)
                else -> {
                    val yesNoAnswer = if (oldCardAction.cards[0].cost > 0 || computer.wantsCoppers()) {
                        "yes"
                    } else {
                        "no"
                    }
                    CardActionHandler.handleSubmittedCardAction(game, player, null!!, yesNoAnswer, null, -1)
                }
            }
            "Mendicant" -> {
                val cardToGain = computer.getHighestCostCard(oldCardAction.cards)
                val cardIds = ArrayList<Int>()
                cardIds.add(cardToGain!!.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
            "Scriptorium" -> {
                //todo determine best action to discard
                val cardToDiscard = computer.getLowestCostCard(oldCardAction.cards)
                val cardIds = ArrayList<Int>()
                cardIds.add(cardToDiscard!!.cardId)
                CardActionHandler.handleSubmittedCardAction(game, player, cardIds, null, null, -1)
            }
        }
    }
}
